flink-ml/flink-ml-api/src/main/java/org/apache/flink/ml/api/core/PipelineStage.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.ml.api.core;
+
+import org.apache.flink.ml.api.misc.param.ParamInfo;
+import org.apache.flink.ml.api.misc.param.WithParams;
+import org.apache.flink.ml.api.misc.persist.Persistable;
+import org.apache.flink.ml.util.param.ExtractParamInfosUtil;
+import org.apache.flink.ml.util.persist.MLStageFactory;
+
+import com.google.gson.Gson;
+import com.google.gson.JsonObject;
+
+import java.io.Serializable;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Base class for a stage in a pipeline. The interface is only a concept, and does not have any
+ * actual functionality. Its subclasses must be either Estimator or Transformer. No other classes
+ * should inherit this interface directly.
+ *
+ * <p>Each pipeline stage is with parameters and meanwhile persistable.
+ *
+ * @param <T> The class type of the PipelineStage implementation itself, used by {@link
+ *            org.apache.flink.ml.api.misc.param.WithParams}
+ * @see WithParams
+ */
+interface PipelineStage<T extends PipelineStage<T>> extends WithParams<T>, Serializable,
+		Persistable {
+
+	default String toJson() {
```

Comment:
Do you mean just using these names or serializing to a byte array like most of serialization?
We hope the serialized String is human-readable and end-users can easily understand and edit it. This makes it possible to change some params and rerun a pipeline without recompiling the code. Json seems to be a standard and friendly way to express the params of a stage. 
On the other hand, className is needed when restoring. It's now appended to the json with signWithClass() method in PipelineStageFactory. It's hard to provide such tool method if the serialized String is not a standard json map format.
That's why we explicitly use 'json' in the method name and why we use json as the serialize result. But we do wonder whether it's good to use json in serializing. What do you think?

Reply:
Yes, this is also my confusion I think. let me rephrase it differently:
1. if this is only for human readability, why do we need a `toJson` and `fromJson`? shouldn't the `toJson` be enough? 
2. I can see that parameter editing is a very tunable part for ML users. I am not expert but do we usually package param literals (e.g. the values) and models together? can we load them differently? e.g. load the param json and the model json separately. 
3. with my personal preference, if it can be a choice for user (json or yaml, or other) that would be even better.

I think one large question I have is whether we use this for serving purpose. if we are not using this for serving purpose. I am actually fine with Json or Yaml.

Reply:
+1 to rename methods for serialization/deserialization model stages.
If we would want to support other formats for e.g. PMML for serialization/deserialization, then toJson/fromJson doesn't match to logic in this case.

I suggest move serialization/deserialization implementation of logic to separate object from direct implementation in this interface - and inject serializer/deserializer to pipeline object.
I mean next
```java
class ExampleTransformer implements Transformer<ExampleTransformer>{
   .... 
   public ExampleTransformer(PipelineSerializer implementationForSpecificFormat){
            this.serializer=implementationForSpecificFormat;
....
    }
....
   public String serialization() {    // renamed version of toJson()
            return this.serializer.serialize(this);
   }
..... //some approach with deserializer
}
```
or move out serialization and deserialization methods from Pipeline objects at all.
```java
Transformer example = new ExampleTransformer();
//actions with ExampleTransformer
String serializedPipeline = new PipelineSerializer("supported_format_name_like_json").serialize(example);
```
I think this approach is more flexible for support different formats for serialization\deserialization pipelines and this the changes we can do in the next PR if community will approve this propolsal.

Reply:
@Rongrong @ex00 @blublinsky
Thank you all for so much concern.

I think most of your questions can be concluded as three points:
1.Why use JSON rather than standard model languages like PMML? What if users want to use other format?
2.How to serve a model?

In fact the Persistable is designed to be used inside FlinkML only. It's not only the storage format but also a part of user interface, like SQL. We tried to design a SQL-like language to describe pipeline stages, but discarded since it's requiring users learn new things, which may be a disadvantage at the beginning of FlinkML. Though the json may be hard to write from zero, it is friendly enough to read and edit. We tried to support these separately but found that if the representation is defined reversible, it can meet both requirements.

It's also planned to support standard model languages. Let's take PMML as the example. Since it is not reversible once a model is converted to PMML, and may be hard to describe all kinds of pipeline stages including pipelines, we didn't choose it as the unified storage format, but to provide a PMMLExportable interface. Models can implement it then users could convert their trained model to a PMML-format string or file.
As for serving, it's supposed that Model.transform() does exactly what serving does. This means a user can easily serve their model trained by FlinkML, just save the model json then use it in his serving job. It's also possible to use the trained model outside Flink if the model implements PMMLExportable. We can also provide PMMLPredictor(extends Transformer) to serve an external PMML model. I also tried implement FLIP-23 proposed by blublinsky on top of this API. I put metadata of the updating model queue(like kafka) in the params then acquire the source in Model.transform(). Though the implementation is quite complex, it's universal and easy to use for users. We can do more to integrate FLIP-39 and FLIP-23 and provide better experience for users.

This is what we thought about, and I hope it solve your problems. I would appreciate you letting me know if you have any question or think there's a mistake.

Reply:
There are a few questions here:
1. I think ML and serving will require 2 separate instances of Flink. ML is mostly batch, while Serving can be either batch or streaming. I think such separation should be very explicit in the overall design. Although theoretically it is possible to use ML to update the model directly, such a use case is not going to be a dominant one.
2. If we agree on this, there are 2 reasons to export the model - here model is really a pipeline. It is used for continue learning and serving. Continue learning, in this case means that I have new data and I want to start from where I left off parameters wise. For this, current persistence implementation might be fine. For serving, it is probably not. You want to to make sure that serving can be done either in Flink or in a standalone Java app, for example. Tensorflow is trying to use the same format for both, but I do not see a big problem separating the two. For serving you might want to try to stick with one of the existing standards to avoid writing your own evaluator.
3. For export, I like the separation of the exporter from pipelines (see https://openscoring.io/blog/2018/07/09/converting_sparkml_pipeline_pmml/). In this case you can roll out additional exporters without modifying the existing code. Actually using the same approach for persistence allows for more flexibility as well

Reply:
@blublinsky
Thanks for your feedback. Your raised a few good points and suggestions, on persistence and model serving. Here are my thoughts.

First of all, I'd like to highlight that the current persistence system is mainly used inside Flink only, and export will be provided in the future as a utility interface.

--------------
I suppose that the persistence system should use a reversible format that can fully express a PipelineStage(including Pipeline) and can be used to restore the stage, whatever the stage is. Saved model can be used for serving or as an initial model of another estimator. Saved estimator may be used in tuning, where users only want to modify some params and rerun the same estimator.

Some 3rd-party system may also use the persistence result. The jpmml-sparkml you mentioned reads a saved PipelineModel and convert it to a PMML, where the saved PipelineModel is consist of saved stages files, and how these files are written is completely defined by its stage class itself. Exporting a pipeline as PMML and saving a pipeline are independent in this case.

In my point of view, the persistence is neither export nor checkpoint. It's more like a JobGraph or SQL, which describes the procedure of pipeline. Besides that, we also tried to find a way like SQL to describe the pipeline, but it's hard to design a new SQL-like language that everyone understands and accepts. JSON is finally chosen.

It's true that the JSON may be too complex to read if everything in a large pipeline put in it. In fact we suppose that the JSON contains only the params set by users, and large models should be saved separately and only the meta is in the JSON. Since where to save the model is also one of the required configurations of estimators, which is usually used to create a Sink, it should be quite natural to put in the params. Of course, utility interfaces and methods like "WithExternalModelFile" would be provided in future to make the implementation easier.

The persisted model JSON can be used to restore the pipeline and serve in Flink. All these should be really easy for a Flink user. If one want to use the trained model in an external serving system that support only standard model format like PMML, he should "export" the model as PMML before using it, and it's not the job persistence should do. For instance, we can implement serving transformers such as PMMLServingTransformer to serve external standard models.

I hope the above has explained “persistence". Next I will try to explain “export", which is not initially included in this PR.

---------------
I really appreciated your 3rd point. Initially, We planned to provide "Exportable" interfaces like PMMLExportable. Model may extend it and provide exportPMML() method to users. It's hard to give a unified implementation for every models, so the interface will be implemented by each model. Exportable requires the model implement it, but it would be hard to extend a model to support exporting without modifying the code, and would be even worst if multiple “export” is going to be supported.

Your 3rd suggestion is a perfect solution to address the above issue, we can implement an exporter utility, which is completely independent from the model. We may first introduce a basic Exporter<M extends Model> interface. And library developers or any one who wants to extend a library can implement it. Users need to create a new exporter instance then call exporter.export(model) to use it as needed. This should be more flexible and also easy to use. 

We can add “exporter” interface in the next PRs after the api and some initial estimators/transformers are merged. I would really appreciate if you can join the discussion when we design it at that time.

-----------------

Thanks again for your feedback. Please let me know if I have addressed all your comments.

Reply:
Thanks for all the comments. It triggers a broad and deep thinking on this interface and the future plan to export/load the model in different formats. I have offline discussed this with @c4emmmm. The final plan looks good to me. @blublinsky, @walterddr, @ex00 , please let us know if you have any further question on the solution proposed by @c4emmmm.

Reply:
+1 to expose more extendable interface in the next PR. 

I am still not fully convince that "toJson" addresses all of the concerns that we raise, especially once @blublinsky raised. but I also have a strong feeling that this can be address in future PR. 
One fundamental supporting factor I am on the "toJson" interface is that: a readable/editable format where a model can be bi-directionally converted in and out of FlinkML is 100% a need in multiple scenarios. for this PR purpose I don't think we need to focus on addressing "all" scenarios. what do you guys think?

IMO the follow up can be on "export", "serving", or even "standardizing" the model

Reply:
Yes @walterddr. "toJson" interface makes it easier to read/edit a persisted pipeline. We can continue to discuss/work on the design of model export/load in the next PR. And it will make the discussion much easier when there is a PR for model export ready for review.
I can understand why the persistable interface introduces a lot confusions here. I talked with @c4emmmm offline again, and here are our decisions:
1. delete persistable interface (it is a common interface, but currently only intend for the persistence of pipelineStage, which does not quite make sense), and leave the implementation of `toJson` and `loadJson` just in PipelineStage.
2. delete the PipelineStageFactory interface as well as the `stageMap` in `toJson` and `loadJson`, as this contract is not ensured from interface.

Reply:
Thanks guys, this plan sounds good for me. Could you create subtask for it in FLINK-12470?

Reply:
@ex00 @walterddr @blublinsky
You may have noticed that we send a mail to flink-dev bringing up a discussion about export and import. Since we have discussed here in the pr, I think maybe you will be interested in the topic. We summarized some thoughts on this in this doc together with some examples:

https://docs.google.com/document/d/1B84b-1CvOXtwWQ6_tQyiaHwnSeiRqh-V96Or8uHqCp8/edit?usp=sharing

Please take a look if you have time.